### PR TITLE
Fixes #10547 - Cannot customize Executor on WebSocketClient.

### LINKED ISSUE #10547: Cannot customize Executor on WebSocketClient
**Jetty version(s)**
10.0.16

**Jetty Environment**
n/a

**Java version/vendor** `(use: java -version)`
All

**OS type/version**
All

**Description**
If a developer wants to customize the Executor of the WebSocketClient, it is not possible, as the internal final WebSocketComponents is initialized with a new Executor from scratch, not using the one from the HttpClient (or other common / shared locations)

### PR DESCRIPTION
**Pull request by sbordet**
The HttpClient and WebSocketComponents will now try to share as many components as possible.

### CODE REVIEW

File: jetty-websocket/websocket-core-client/src/main/java/org/eclipse/jetty/websocket/core/client/WebSocketCoreClient.java [hunk 1]
```diff
@@ -52,16 +52,25 @@ public WebSocketCoreClient(WebSocketComponents webSocketComponents)
...
-            httpClient = Objects.requireNonNull(HttpClientProvider.get());
-        if (httpClient.getExecutor() == null)
-            httpClient.setExecutor(webSocketComponents.getExecutor());
-
-        this.classLoader = Thread.currentThread().getContextClassLoader();
-        this.httpClient = httpClient;
-        this.components = webSocketComponents;
-        addBean(httpClient);
-        addBean(webSocketComponents);
+        client = Objects.requireNonNullElse(httpClient, HttpClientProvider.get());
```
Comment (lorban) [inline jetty-websocket/websocket-core-client/src/main/java/org/eclipse/jetty/websocket/core/client/WebSocketCoreClient.java:55]: There should be tests for `client.isStarted()`, `!client.isStarted()` and `webSocketComponents == null`, `webSocketComponents != null` code paths to make assert that the buffer pool and the executor are shared as expected.

**Review by lorban — CHANGES_REQUESTED** [review]

File: jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/WebSocketClient.java [hunk 2]
```diff
@@ -64,61 +63,27 @@ public class WebSocketClient extends ContainerLifeCycle implements WebSocketPoli
...
-     *
-     * <p>
-     *     HTTP behaviors of the WebSocket upgrade are taken from the {@link HttpClient} configuration.
-     *     WebSocket behaviors are taken from the {@link WebSocketComponents} configuration.
-     * </p>
-     *
-     * @param httpClient the HttpClient to use for the HTTP behaviors of WebSocket upgrade
-     * @param webSocketComponents the WebSocketComponents to use for WebSocket behaviors
-     */
-    public WebSocketClient(HttpClient httpClient, WebSocketComponents webSocketComponents)
```
Comment (joakime) [inline jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/WebSocketClient.java:116]: Why did you remove this constructor?
It was added specifically to support this use case.

Comment (sbordet) [reply jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/WebSocketClient.java:116]: @joakime read the comment in the issue. The addition of that constructor was wrong.

File: jetty-websocket/websocket-core-client/src/main/java/org/eclipse/jetty/websocket/core/client/WebSocketCoreClient.java (hunk 1 above)
Comment (sbordet) [reply jetty-websocket/websocket-core-client/src/main/java/org/eclipse/jetty/websocket/core/client/WebSocketCoreClient.java:55]: @lorban I added tests only for the started/!started cases, as you cannot control WebSocketComponents (it's always null).

Comment (lorban) [reply jetty-websocket/websocket-core-client/src/main/java/org/eclipse/jetty/websocket/core/client/WebSocketCoreClient.java:55]: What if you call this constructor with a null `WebSocketComponents`?

Comment (sbordet) [reply jetty-websocket/websocket-core-client/src/main/java/org/eclipse/jetty/websocket/core/client/WebSocketCoreClient.java:55]: @lorban it'll be created.

**Review by lorban — APPROVED** [review]

**Review by joakime — APPROVED** [review]
